lib/plug/cowboy/drainer.ex

```diff
@@ -0,0 +1,79 @@
+defmodule Plug.Cowboy.Drainer do
+  @moduledoc """
+  TODO: description
+  """
+  use GenServer
+
+  @doc """
+  TODO: description
+
+  ## Options
+
+    * `:id` - The ID for the process.
+      Defaults to `Plug.Cowboy.Drainer`.
+
+    * `:shutdown` - How long to wait for connections to drain.
+      Defaults to 5000ms.
+
+    * `:refs` - The cowboy listener refs to drain connections for.
+      Defaults to all cowboy listeners.
+
+    * `:drain_check_interval` - How frequently to check if a listener's
+      connections have been drained.
+      Defaults to 1000ms.
+  """
+  @spec child_spec(opts :: Keyword.t()) :: Supervisor.child_spec()
+  def child_spec(opts) when is_list(opts) do
+    {spec_opts, opts} = Keyword.split(opts, [:id, :shutdown])
+
+    Supervisor.child_spec(
+      %{
+        id: __MODULE__,
+        start: {__MODULE__, :start_link, [opts]},
+        type: :worker
+      },
+      spec_opts
+    )
+  end
+
+  @doc false
+  def start_link(opts) do
+    GenServer.start_link(__MODULE__, opts)
+  end
+
+  @doc false
+  @impl true
+  def init(opts) do
+    Process.flag(:trap_exit, true)
+    {:ok, opts}
+  end
+
+  @doc false
+  @impl true
+  def terminate(_reason, opts) do
+    opts
+    |> Keyword.get(:refs)
```

Comment:
Should we make drain all an opt-in feature? For example: `refs: :all`? Instead of relying on `nil`?

Reply:
Seems reasonable

Reply:
Perhaps it should become a proper arg since in that case it'd no longer be optional.

Reply:
Btw, we should check this when building the child_spec or on start_link.